File: Week 1 C/credit.c
// Implement Luhn's Algorithm to check if a credit crad is valid

#include <stdio.h>
#include <cs50.h>
#include <string.h>

long get_positive_long(string prompt);
int get_digit(long n);

int main(void)
{
    long credit = get_positive_long("Number: ");
    int N = get_digit(credit);  
    
    // Generate an array of credit card numbers (integers)
    // the last credit number is the first integer in the array
    int digit_arr[N];
    for (int i = 0; i < N; i++)
    {
        digit_arr[i] = credit % 10;
        credit = (credit - digit_arr[i]) / 10;
    }
    
    // sum_odd is sum of odd digits in credit card number (even position in the array)
    int sum_odd = 0;
    for (int i = 0; i < N; i += 2)
    {
        sum_odd += digit_arr[i];
    }

    // sum_even is sum of (2 * even digits) in credit card number (odd position in the array)
    int sum_even = 0;
    for (int i = 1; i < N; i += 2)
    {
        if (digit_arr[i] < 5)   // Only 1-digit
        {
            sum_even += 2 * digit_arr[i];
        }
        else  // 2-digit needs to be split and add up (e.g. 12 => 1 + 2)
        {
            sum_even += 1 + (2 * digit_arr[i]) % 10;
        }
    }

    // Check if credit card is valid
    if ((sum_odd + sum_even) % 10 == 0)
    {
        // Check if number is American Express
        // Length is 15 and start with 34 or 37
        if (N == 15 && digit_arr[14] == 3)
        {
            if (digit_arr[13] == 4 || digit_arr[13] == 7)
            {
                printf("AMEX\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        // Check if number is Visa
        // Length is 13 or 16 and start with 4
        else if (N == 16 && digit_arr[15] == 4)
        {
            printf("VISA\n");
        }
        else if (N == 13 && digit_arr[12] == 4)
        {
            printf("VISA\n");
        }
        // Check if number is MasterCard
        // Length is 16 and start with 51, 52, 53, 54, or 55
        else if (N == 16 && digit_arr[15] == 5)
        {
            if (digit_arr[14] == 1 || digit_arr[14] == 2 || digit_arr[14] == 3 || digit_arr[14] == 4 || digit_arr[14] == 5)
            {
                printf("MASTERCARD\n");
            }
            else
            {
                printf("INVALID\n");
            }
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}


// Get a non-negative long from user
// Keep asking if user did not input the right number
long get_positive_long(string prompt)
{
    long n;
    do
    {
        n = get_long("%s", prompt);
    }
    while (n < 0);
    return n;
}

// How many digit does the long number have?
int get_digit(long n)
{
    int count = 0;
    
    do
    {
        n /= 10;
        count ++;
    }
    while (n != 0);
    return count;
}

File: Week 6 Python/credit.py
from cs50 import get_string

credit = get_string("Number: ")
N = len(credit)
# Total of odd digits
if N % 2 == 1:
    # odd digits are in index 0, 2, 4, ...
    # even digits are in index 1, 3, 5, ...
    m = 0
    n = 1
# Total of even digits
else:
    # odd digits are in index 1, 3, 5, ...
    # even digits are in index 0, 2, 4, ...
    m = 1
    n = 0

sum_odd = 0
# sum_odd is sum of odd digits in credit card number (even position in the array)
for i in range(m, N, 2):
    sum_odd += int(credit[i])

sum_even = 0
# sum_even is sum of (2 * even digits) in credit card number (odd position in the array)
for i in range(n, N, 2):
    if int(credit[i]) < 5:
        sum_even += 2 * int(credit[i])
    else:
        sum_even += 1 + (2 * int(credit[i])) % 10

# Check if credit card is valid
if (sum_odd + sum_even) % 10 == 0:
    # Check if number is American Express
    # Length is 15 and start with 34 or 37
    if N == 15 and int(credit[0]) == 3 and int(credit[1]) in [4, 7]:
        print("AMEX")
    # Check if number is Visa
    # Length is 13 or 16 and start with 4
    elif N in [13, 16] and int(credit[0]) == 4:
        print("VISA")
    # Check if number is MasterCard
    # Length is 16 and start with 51, 52, 53, 54, or 55
    elif N == 16 and int(credit[0]) == 5 and int(credit[1]) in [1, 2, 3, 4, 5]:
        print("MASTERCARD")
    else:
        print("INVALID")
else:
    print("INVALID")
